### PR TITLE
Editable hostname

### PR DESCRIPTION
**Pull request by KeironO**
This PR allows users to change the hostname of their system from within the About plug, as described in Issue https://github.com/elementary/switchboard-plug-about/issues/90.

change

I've implemented a ```clean_hostname``` method that can be extended to ensure that hostnames are safe.

Best wishes,

Keiron

### CODE REVIEW

File: src/Plug.vala [hunk 1]
```diff
@@ -195,6 +195,7 @@ public class About.Plug : Switchboard.Plug {
             }
         });
 
+
```
Comment (danrabbit) [inline src/Plug.vala:198]: Looks like some extra white space got added here

File: src/Views/HardwareView.vala [hunk 2]
```diff
@@ -1,23 +1,23 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA
+ */
+
+public class About.HardwareView: Gtk.Grid {
```
Comment (danrabbit) [inline src/Views/HardwareView.vala:20]: Can you revert all these whitespace changes please? :)

**Comment by KeironO** [comment]
Apologies, I think I may had inadvertently done something in Code without realising it. This should fix the problem.

**Comment by danrabbit** [comment]
That's starting to look a lot better, but if you check the diff in GitHub you can see a lot of places where code style was changed still.

Also, it looks like your branch has merge conflicts with Master that need to be resolved

Thanks for continuing to work on this!

**Comment by KeironO** [comment]
So it looks like there has been an update to the version of glib this uses, I'm running eOS 5.1.6 which uses 2.56 - so I'm not sure what to do now.

**Comment by danrabbit** [comment]
@KeironO Ah right. For now you could build on Ubuntu Focal, or wait until we have developer preview images available for 6.0

**Comment by KeironO** [comment]
I'll get a focal VM set up, and get this sorted when I'm free.

Any ETA on when the developer images are going to be available, or is it a case of "when they're ready"?

**Comment by danrabbit** [comment]
Working on it as fast as we can :)

File: src/Views/HardwareView.vala [hunk 3]
```diff
@@ -188,17 +220,41 @@ public class About.HardwareView : Gtk.Grid {
         }
 
         // Memory
-        memory = GLib.format_size (get_mem_info ());
+        memory = GLib.format_size ( get_mem_info ());
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:223]: Remove added space

```suggestion
        memory = GLib.format_size (get_mem_info ());
```

File: src/Views/HardwareView.vala [hunk 4]
```diff
@@ -188,17 +220,41 @@ public class About.HardwareView : Gtk.Grid {
...
         graphics = clean_graphics_name (session_manager.renderer);
 
         // Hard Drive
         var file_root = GLib.File.new_for_path ("/");
         string storage_capacity = "";
         try {
-            var info = file_root.query_filesystem_info (GLib.FileAttribute.FILESYSTEM_SIZE, null);
-            storage_capacity = GLib.format_size (info.get_attribute_uint64 (GLib.FileAttribute.FILESYSTEM_SIZE));
+            var info = file_root.query_filesystem_info(GLib.FileAttribute.FILESYSTEM_SIZE, null);
+            storage_capacity = GLib.format_size(info.get_attribute_uint64(GLib.FileAttribute.FILESYSTEM_SIZE));
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:257]: Revert code style changes

```suggestion
            var info = file_root.query_filesystem_info (GLib.FileAttribute.FILESYSTEM_SIZE, null);
            storage_capacity = GLib.format_size (info.get_attribute_uint64 (GLib.FileAttribute.FILESYSTEM_SIZE));
```

File: src/Views/HardwareView.vala [hunk 5]
```diff
@@ -77,20 +90,22 @@ public class About.HardwareView : Gtk.Grid {
...
+        attach (hostname_info, 1, 4, 1, 1);
+        attach (hostname_label, 0, 4);
+        attach (processor_info, 0, 5, 2, 1);
+        attach (graphics_info, 0, 6, 2, 1);
+        attach (memory_info, 0, 7, 2, 1);
+        attach (hdd_info, 0, 8, 2, 1);
 
         if (oem_enabled) {
-            var fileicon = new FileIcon (File.new_for_path (manufacturer_icon_path));
+            var fileicon = new FileIcon(File.new_for_path (manufacturer_icon_path));
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:108]: Revert code style change

```suggestion
            var fileicon = new FileIcon (File.new_for_path (manufacturer_icon_path));
```

File: src/Views/HardwareView.vala [hunk 6]
```diff
@@ -105,38 +120,55 @@ public class About.HardwareView : Gtk.Grid {
...
-                attach (manufacturer_website_info, 0, 7, 2, 1);
+                attach (manufacturer_website_info, 0, 9, 2, 1);
             }
         } else {
             attach (product_name_info, 0, 1, 2, 1);
         }
     }
 
-    private void fetch_hardware_info () {
+    private string clean_hostname(string hostname) {
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:147]: Code style: space before paren

```suggestion
    private string clean_hostname (string hostname) {
```

File: src/Views/HardwareView.vala [hunk 7]
```diff
@@ -105,38 +120,55 @@ public class About.HardwareView : Gtk.Grid {
...
+        string cleaned_hostname = hostname.replace (" ", "-");
+        return cleaned_hostname;
+    }
+
+    private void change_hostname(string hostname) {
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:152]: Code style: space before paren

```suggestion
    private void change_hostname (string hostname) {
```

File: src/Views/HardwareView.vala [hunk 8]
```diff
@@ -105,38 +120,55 @@ public class About.HardwareView : Gtk.Grid {
...
+        if (hostname.length == 0) {
+            hostname = "localhost";
+        }
+
+        try {
+            Process.spawn_command_line_async ("hostnamectl set-hostname %s".printf(hostname.to_ascii ()));
+        } catch (SpawnError e) {
+            warning (e.message);
+        }
+
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:162]: Unneeded blank line

```suggestion
```

File: src/Views/HardwareView.vala [hunk 9]
```diff
@@ -105,38 +120,55 @@ public class About.HardwareView : Gtk.Grid {
...
+    }
+
+    private void fetch_hardware_info() {
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:165]: Code style: space before paren

```suggestion
    private void fetch_hardware_info () {
```

File: src/Views/HardwareView.vala [hunk 10]
```diff
@@ -105,38 +120,55 @@ public class About.HardwareView : Gtk.Grid {
...
         // Processor
         var cpu_file = File.new_for_path ("/proc/cpuinfo");
-        uint cores = 0U;
+        uint cores = 0;
         bool cores_found = false;
         try {
-            var dis = new DataInputStream (cpu_file.read ());
+            var dis = new DataInputStream (cpu_file.read());
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:171]: Code style: space before paren

```suggestion
            var dis = new DataInputStream (cpu_file.read());
```

File: src/Views/HardwareView.vala [hunk 11]
```diff
@@ -188,17 +220,41 @@ public class About.HardwareView : Gtk.Grid {
...
 
         // Graphics
+        try {
+            Process.spawn_command_line_sync ("lspci", out graphics);
+
+            if ("VGA" in graphics) { //VGA-keyword indicates graphics-line
+                string[] lines = graphics.split ("\n");
+                graphics = "";
+
+                foreach(var s in lines) {
```
Comment (cassidyjames) [inline src/Views/HardwareView.vala:233]: Code style: space before paren

```suggestion
                foreach (var s in lines) {
```

File: src/Views/HardwareView.vala [hunk 12]
```diff
@@ -20,7 +20,7 @@
 * Boston, MA 02110-1301 USA
 */
 
-public class About.HardwareView : Gtk.Grid {
+public class About.HardwareView: Gtk.Grid {
```
Comment (tintou) [inline src/Views/HardwareView.vala:23]: the previous version was the right one

File: src/Views/HardwareView.vala [hunk 13]
```diff
@@ -58,10 +58,23 @@ public class About.HardwareView : Gtk.Grid {
...
         product_name_info.ellipsize = Pango.EllipsizeMode.END;
         product_name_info.get_style_context ().add_class ("h2");
-        product_name_info.set_selectable (true);
+        product_name_info.set_selectable(true);
+
+        var hostname_label = new Gtk.Label (_("Hostname:"));
+        hostname_label.halign = Gtk.Align.END;
+
+        var hostname_info = new Gtk.Entry ();
+        hostname_info.set_text(Environment.get_host_name ());
```
Comment (tintou) [inline src/Views/HardwareView.vala:70]:
```suggestion
        hostname_info.set_text (Environment.get_host_name ());
```

File: src/Views/HardwareView.vala [hunk 14]
```diff
@@ -77,17 +90,19 @@ public class About.HardwareView : Gtk.Grid {
         graphics_info.justify = Gtk.Justification.CENTER;
         graphics_info.set_selectable (true);
 
-        var hdd_info = new Gtk.Label (hdd);
+        var hdd_info = new Gtk.Label(hdd);
```
Comment (tintou) [inline src/Views/HardwareView.vala:93]:
```suggestion
        var hdd_info = new Gtk.Label (hdd);
```

File: src/Views/HardwareView.vala [hunk 15]
```diff
@@ -105,35 +120,51 @@ public class About.HardwareView : Gtk.Grid {
...
+        return cleaned_hostname;
+    }
+
+    private void change_hostname (string hostname) {
+        if (hostname.length == 0) {
+            hostname = "localhost";
+        }
+
+        try {
+            Process.spawn_command_line_async ("hostnamectl set-hostname %s".printf(hostname.to_ascii ()));
```
Comment (tintou) [inline src/Views/HardwareView.vala:158]: That's not the right way to do it. The right way is to use the `org.freedesktop.hostname1` D-Bus interface on the system bus. This also allows to make sure that the user actually has the right to change the hostname (which should be shown in the interface too)

File: src/Views/HardwareView.vala [hunk 16]
```diff
@@ -191,6 +222,30 @@ public class About.HardwareView : Gtk.Grid {
         memory = GLib.format_size (get_mem_info ());
 
         // Graphics
+        try {
+            Process.spawn_command_line_sync ("lspci", out graphics);
```
Comment (tintou) [inline src/Views/HardwareView.vala:226]: There should be another way of doing this as this is not portable across platforms (there is no lspci on ARM for instance)

**Comment by cassidyjames** [comment]
Thanks again for your work on this. However, after discussing with the UX team, we're closing in favor of https://github.com/elementary/switchboard-plug-sharing/issues/53

**Pull request closed without merging**